src/wallet/wallet.cpp

```diff
@@ -1419,52 +1416,214 @@ ZTXOSelector CWallet::LegacyTransparentZTXOSelector() {
+                    // the payment source account.
+                    return self->GetUnifiedAccountId(meta.value().first) == std::optional(acct.GetAccountId());
+                } else {
+                    return false;
+                }
+            }
+            return false;
+        }
+    }, selector.GetPattern());
+}
+
+SpendableInputs CWallet::FindSpendableInputs(
+        ZTXOSelector selector,
+        bool allowTransparentCoinbase,
+        uint32_t minDepth) const {
+    AssertLockHeld(cs_main);
+    AssertLockHeld(cs_wallet);
+
+    KeyIO keyIO(Params());
+
+    bool selectTransparent{selector.SelectsTransparent()};
+    bool selectSprout{selector.SelectsSprout()};
+    bool selectSapling{selector.SelectsSapling()};
+
+    SpendableInputs unspent;
+    for (auto const& [wtxid, wtx] : mapWallet) {
+        bool isCoinbase = wtx.IsCoinBase();
+        auto nDepth = wtx.GetDepthInMainChain();
+
+        // Filter the transactions before checking for coins
+        if (!CheckFinalTx(wtx)) continue;
+        if (nDepth < minDepth) continue;
+
+        if (selectTransparent &&
+            // skip transparent utxo selection if coinbase spend restrictions are not met
+            (!isCoinbase || (allowTransparentCoinbase && wtx.GetBlocksToMaturity() <= 0))) {
+
+            for (int i = 0; i < wtx.vout.size(); i++) {
+                const auto& output = wtx.vout[i];
+                isminetype mine = IsMine(output);
+
+                // skip spent utxos
+                if (IsSpent(wtxid, i)) continue;
+                // skip utxos that do not belong to the wallet
+                if (mine == ISMINE_NO || ((mine & ISMINE_SPENDABLE) == ISMINE_NO)) continue;
+                // skip locked utxos
+                if (IsLockedCoin(wtxid, i)) continue;
+                // skip zero-valued utxos
+                if (output.nValue == 0) continue;
+
+                // check to see if the coin conforms to the payment source
+                CTxDestination address;
+                bool isSelectable =
+                    ExtractDestination(output.scriptPubKey, address) &&
+                    this->SelectorMatchesAddress(selector, address);
+                if (isSelectable) {
+                    unspent.utxos.push_back(COutput(&wtx, i, nDepth, true, isCoinbase));
+                }
+            }
+        }
+
+        if (selectSprout) {
+            for (auto const& [jsop, nd] : wtx.mapSproutNoteData) {
+                SproutPaymentAddress pa = nd.address;
+
+                // skip note which has been spent
+                if (nd.nullifier.has_value() && IsSproutSpent(nd.nullifier.value())) continue;
+                // skip notes which don't match the source
+                if (!this->SelectorMatchesAddress(selector, pa)) continue;
+                // skip notes for which we don't have the spending key
+                if (!this->HaveSproutSpendingKey(pa)) continue;
```

Comment:
Shouldn't this respect the selector? i.e. don't require the spending key if the selector was configured that way?

Reply:
Ah, that's a good point, now that the selector carries that flag, I should actually respect it for all of the different input types.